### ISSUE TITLE
Add COM1 and COM2 Relay Control for Indoor Station

### ISSUE THREAD

**Issue by ukiews** [judge/homebridge-hikconnect#25]
For Indoor Stations like the DS-KH6350 there are two relays on board (COM1 and COM2).
Page 10 here: https://assets.hikvision.com/prd/public/all/doc/m000056209/UD30790B-A_Network-Indoor-Station_Operation-Guide_V2.2.2_20240819.pdf

Could you please add these?

The plugin already pulls the paired door station locks from the indoor station, but adding the two relays would allow for full control.

![IMG_5401](https://github.com/user-attachments/assets/364aafcf-24b6-4a90-9a46-3ea1f5a90376)

**Comment by ukiews** [comment]
@judge any chance this could be added?

**Comment by judge** [comment]
Hello,
Currently I'm busy with life stuff, I'll check this soon, sorry about that!
In the meantime, could you please help me with the payload you receive from the station? I don't have these relays in my system so I cannot test it. There is a guide about how to send me the response: https://github.com/judge/homebridge-hikconnect/blob/main/DEBUGGING.md

**Comment by ukiews** [comment]
Thank you for getting back.
Looks like Step 2 in Debugging instructions does not work.
Even though I am logged into Hik-Connect, the url kicks back a 401 error: "{"meta":{"code":401,"message":"Unauthorized"}}"
Do I need a developer account to accomplish this since the link is "api.hik-connect.com..."?

**Comment by ukiews** [comment]
@judge I opened a dev account at tpp.hikvision.com, but I still get the same 401 error when clicking your link.

**Comment by judge** [comment]
I checked it as well and it seems the method I described in the docs is not working anymore. You need to craft the requests by hand. :( I'll publish a new version of the plugin to help with debugging, I have an idea.

**Comment by judge** [comment]
I published a new version with debug logging. Could you please follow the updated https://github.com/judge/homebridge-hikconnect/blob/main/DEBUGGING.md docs?

**Comment by ukiews** [comment]
Log sent!
There seem to be nested arrays marked as '[Object]' that don't fully populate. Not sure if that is relevant.

**Comment by ukiews** [comment]
@judge I was able to do a bit of reverse engineering by monitoring the traffic from my iPhone and the Hik-Connect app to control the Com1 port:

```
    "onUrl": {
        "url": "https://apiius.hik-connect.com/v3/userdevices/v1/isapi",
        "method": "POST",
        "headers": {
            "Content-Type": "application/x-www-form-urlencoded",
            "featureCode": "xxx",
            "appId": "NewHikConnect",
            "sessionId": "xxx"
        },
        "body": "apiData=PUT%20/ISAPI/SecurityCP/control/outputs/0?format%3Djson%0D%0A%7B%0A%20%20%22OutputsCtrl%22%20%3A%20%7B%0A%20%20%20%20%22switch%22%20%3A%20%22open%22%0A%20%20%7D%0A%7D&apiKey=100014&channelNo=0&deviceSerial=Q33452178&method=1"
    },
    "offUrl": {
        "url": "https://apiius.hik-connect.com/v3/userdevices/v1/isapi",
        "method": "POST",
        "headers": {
            "Content-Type": "application/x-www-form-urlencoded",
            "featureCode": "xxx",
            "appId": "NewHikConnect",
            "sessionId": "xxx"
        },
        "body": "apiData=PUT%20/ISAPI/SecurityCP/control/outputs/0?format%3Djson%0D%0A%7B%0A%20%20%22OutputsCtrl%22%20%3A%20%7B%0A%20%20%20%20%22switch%22%20%3A%20%22close%22%0A%20%20%7D%0A%7D&apiKey=100014&channelNo=0&deviceSerial=Q33452178&method=1"
    }
```
Would this help?
Only trouble is that I could not use regular digest authentication for some reason.

**Comment by judge** [comment]
Hmm it seems this is based on isapi, which is an API interface for hikvision devices. My plugin does not use that, but I also sniffed data between my device and phone app. I cannot enable isapi on my device because I live in a flat and my device is managed by someone else. :(
So to implement something like this into the plugin, I would need to use the isapi but I can't. Do you have local access to you device? If so, there is a forum thread on HomeAssistant forums (https://community.home-assistant.io/t/ds-kd8003-ds-kv8113-ds-kv8213-ds-kv6113-ds-kv8413-and-integration-hikvision-hikconnect-video-intercom-doorbell/238535), the LAN access is based on isapi as far as I know and it supports much more features than my plugin.

**Comment by ukiews** [comment]
> Hmm it seems this is based on isapi, which is an API interface for hikvision devices. My plugin does not use that, but I also sniffed data between my device and phone app. I cannot enable isapi on my device because I live in a flat and my device is managed by someone else. :( So to implement something like this into the plugin, I would need to use the isapi but I can't. Do you have local access to you device? If so, there is a forum thread on HomeAssistant forums (https://community.home-assistant.io/t/ds-kd8003-ds-kv8113-ds-kv8213-ds-kv6113-ds-kv8413-and-integration-hikvision-hikconnect-video-intercom-doorbell/238535), the LAN access is based on isapi as far as I know and it supports much more features than my plugin.

Yes I have looked and interacted there. Unfortunately, for some reason status request through ISAPI does not work using reverse engineering (no documentation provided). Did the log I send help at all?

**Comment by judge** [comment]
Unfortunately, no, I cannot see any info about the relays in that payload. Maybe if I would have a device like that (with relays) I could reverse engineer it, but it seems it's ISAPI only. :(

**Comment by trempetic92** [comment]
> @judge I was able to do a bit of reverse engineering by monitoring the traffic from my iPhone and the Hik-Connect app to control the Com1 port:
> 
> ```
>     "onUrl": {
>         "url": "https://apiius.hik-connect.com/v3/userdevices/v1/isapi",
>         "method": "POST",
>         "headers": {
>             "Content-Type": "application/x-www-form-urlencoded",
>             "featureCode": "xxx",
>             "appId": "NewHikConnect",
>             "sessionId": "xxx"
>         },
>         "body": "apiData=PUT%20/ISAPI/SecurityCP/control/outputs/0?format%3Djson%0D%0A%7B%0A%20%20%22OutputsCtrl%22%20%3A%20%7B%0A%20%20%20%20%22switch%22%20%3A%20%22open%22%0A%20%20%7D%0A%7D&apiKey=100014&channelNo=0&deviceSerial=xxx&method=1"
>     },
>     "offUrl": {
>         "url": "https://apiius.hik-connect.com/v3/userdevices/v1/isapi",
>         "method": "POST",
>         "headers": {
>             "Content-Type": "application/x-www-form-urlencoded",
>             "featureCode": "xxx",
>             "appId": "NewHikConnect",
>             "sessionId": "xxx"
>         },
>         "body": "apiData=PUT%20/ISAPI/SecurityCP/control/outputs/0?format%3Djson%0D%0A%7B%0A%20%20%22OutputsCtrl%22%20%3A%20%7B%0A%20%20%20%20%22switch%22%20%3A%20%22close%22%0A%20%20%7D%0A%7D&apiKey=100014&channelNo=0&deviceSerial=xxx&method=1"
>     }
> ```
> 
> Would this help? Only trouble is that I could not use regular digest authentication for some reason. SessionId expires every 24 hours.

This requires fetching a bearerToken using AppKey and App secret with developer account on hik connect or hik partner pro. Then you can redo requests

**Comment by judge** [comment]
Thanks! The plugin uses the normal end-user hik-connect account, that way the communication is not going through isapi, so it's pretty limited what it can do.
How do you guys get a developer account? When I try to register, I have to register as a company with valid company informations.

**Comment by ukiews** [comment]
It does not seem they do any real verification on company info :)
